src/python/pants/backend/java/tailor.py

```diff
@@ -0,0 +1,71 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import os
+from dataclasses import dataclass
+from typing import Iterable
+
+from pants.backend.java.target_types import JavaLibrary, JavaTestsSources, JunitTests
+from pants.core.goals.tailor import (
+    AllOwnedSources,
+    PutativeTarget,
+    PutativeTargets,
+    PutativeTargetsRequest,
+    group_by_dir,
+)
+from pants.engine.fs import PathGlobs, Paths
+from pants.engine.internals.selectors import Get
+from pants.engine.rules import collect_rules, rule
+from pants.engine.target import Target
+from pants.engine.unions import UnionRule
+from pants.source.filespec import Filespec, matches_filespec
+from pants.util.logging import LogLevel
+
+
+@dataclass(frozen=True)
+class PutativeJavaTargetsRequest(PutativeTargetsRequest):
+    pass
+
+
+def classify_source_files(paths: Iterable[str]) -> dict[type[Target], set[str]]:
+    """Returns a dict of target type -> files that belong to targets of that type."""
+    tests_filespec = Filespec(includes=list(JavaTestsSources.default))
+    test_filenames = set(
+        matches_filespec(tests_filespec, paths=[os.path.basename(path) for path in paths])
+    )
+    test_files = {path for path in paths if os.path.basename(path) in test_filenames}
+    library_files = set(paths) - test_files
```

Comment:
Do we have a preference for directory structure when doing our builds? Is it standard `src/main`/`src/test`, or is it intermingled like we're doing with Python? I'm not 100% certain as to where we want files, and if they're intermingled, this may have the potential to ignore source files that end in `Test.java` -- we could alternatively check to see if `XTest.java` has a corresponding `X.java` to determine if it's a test file (though this is brittle to typos)

Reply:
> Do we have a preference for directory structure when doing our builds?

Not to my knowledge currently. 

> I'm not 100% certain as to where we want files, and if they're intermingled, this may have the potential to ignore source files that end in Test.java

Opened https://github.com/pantsbuild/pants/issues/12792 for one of us to figure out the right answer here. For now, this PR helps reduce my typing in BUILD boilerplate and so is a small step forward.

Reply:
Totally okay for the semantics to change in upcoming PRs given that this backend is highly experimental and not documented.